From 72db8d63ce8896728f94c645bcb3969398822c46 Mon Sep 17 00:00:00 2001
From: SeongJae Park <sj@kernel.org>
Date: Thu, 23 May 2024 15:48:07 -0700
Subject: [PATCH] ==== write-only monitoring ====

Implement new monitoring primitive sets called vaddr-wo and paddr-wo.
Those are same to vaddr and paddr, but use write-protect faults
mechanism for access checking.

DAMON core change
-----------------

Extend DAMON core to provide a function for finding DAMON region of a
given address.  Internally, we can maintain an array for the fast
search.  Each element of the entry contains each DAMON region's
start/end address and a pointer to the region.  The elements are sorted
by the start address.  The array can be populated at last regions
traversal in kdamond.  Then we can use binary search on the array.

Primitive set implementation
----------------------------

On '->prepare_access_checks()', set PTE of the access sampling page as
write-protected.

On write-protect page faults (wp_huge_pud(), wp_huge_pmd(),
do_wp_page()), find the region of the region of faulted page, mark it as
accessed.

On '->check_accesses()', just read the mark.

Set PTE as write-protected on 'prepare_access_checks'

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 damon_meta_changes/qLG8W05V | 0
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 damon_meta_changes/qLG8W05V

diff --git a/damon_meta_changes/qLG8W05V b/damon_meta_changes/qLG8W05V
new file mode 100644
index 000000000000..e69de29bb2d1
-- 
2.39.2

